scipy/optimize/optimize.py

```diff
@@ -2397,20 +2397,84 @@ def bracket(func, xa=0.0, xb=1.0, args=(), grow_limit=110.0, maxiter=1000):
     return xa, xb, xc, fa, fb, fc, funcalls
 
 
-def _linesearch_powell(func, p, xi, tol=1e-3):
+def _m_min(a, b):
+    """ Return min(a, b), support if a or b is None """
+    try:
+        return min(a, b)
+    except TypeError:
+        return a or b
+
+
+def _m_max(a, b):
+    """ Return max(a, b), support if a or b is None """
+    try:
+        return max(a, b)
+    except TypeError:
+        return a or b
+
+
+def _line_for_search(x0, alpha, lower_bound, upper_bound):
+    """
+    x0 is the vector representing the current location (np.array)
+    alpha is the vector representing the direction (np.array)
+
+    ie the direction is along the line from x0 to alpha
+
+    lower_bound is a np.array of the lower bounds for each parameter in x0.
+    upper_bound is a np.array of the upper bounds for each parameter in x0.
+
+    returns (lmin, lmax), the bounds for l such that
+            lower_bound[i] <= x0_i+alpha_i * l <= upper_bound[i]
+        for all i.
+    """
+    # get nonzero indices of alpha so we don't get any zero division errors.
+    nonzero, = alpha.nonzero()
+    lower_bound, upper_bound = lower_bound[nonzero], upper_bound[nonzero]
+    x0, alpha = x0[nonzero], alpha[nonzero]
+    low = (lower_bound - x0) / alpha
+    high = (upper_bound - x0) / alpha
+
+    pos, = np.nonzero(alpha > 0)  # positive indices
+    neg, = np.nonzero(alpha < 0)  # negative indices
+
+    lmin = _m_max(
+        np.max(low[pos]) if pos.shape[0] else None,
+        np.max(high[neg]) if neg.shape[0] else None
+    )
+    lmax = _m_min(
+        np.min(low[neg]) if neg.shape[0] else None,
+        np.min(high[pos]) if pos.shape[0] else None
+    )
+
```

Comment:
I'm trying to avoid the multiple calls to `np.min` and the separate `_m_max`/`_m_min` functions. Would something like:
```
pos = alpha > 0  # positive indices
neg = alpha < 0  # negative indices
lmin = np.max(low*pos + high*neg)
lmax = np.min(low*neg + high*pos)
```
work? Or if that is a little wrong, can you modify the idea to make it work?

Reply:
Is it possible for all of `alpha` to be zero, or is that prevented elsewhere?

Reply:
It is prevented in `_linesearch_powell`, which is the only place that `_line_for_search` is called from.

Reply:
Added your changes, thanks and great idea!